src/fsharp/FSharp.Core/tasks.fs

```diff
@@ -0,0 +1,424 @@
+        sm.ResumptionPoint <- pc
+        let mutable awaiter = awaiter
+        assert (not (isNull awaiter))
+        // Tell the builder to call us again when done.
+        //Console.WriteLine("[{0}] AwaitUnsafeOnCompleted", sm.GetHashCode())
+        sm.MethodBuilder.AwaitUnsafeOnCompleted(&awaiter, &sm)
+
+    let inline SetResult (sm: byref<TaskStateMachineTemplate<'T>>) (v: 'T) = 
+        sm.Result <- v
+
+[<AutoOpen>]
+module TaskHelpers = 
+
+    //let inline unwrapException (agg : AggregateException) =
+    //    let inners = agg.InnerExceptions
+    //    if inners.Count = 1 then inners.[0]
+    //    else agg :> Exception
+
+    /// Used to return a value.
+    let inline ret<'T> (x : 'T) = 
+        Helpers.SetResult __machineAddr<TaskStateMachineTemplate<'T>> x
+        TaskStep<'T, 'T>(true)
+
+    let inline RequireCanBind< ^Priority, ^TaskLike, ^TResult1, 'TResult2, 'TOverall when (^Priority or ^TaskLike): (static member CanBind : ^Priority * ^TaskLike * (^TResult1 -> TaskStep<'TResult2, 'TOverall>) -> TaskStep<'TResult2, 'TOverall>) > (x: ^Priority) (y: ^TaskLike) __expand_continuation = 
+        ((^Priority or ^TaskLike): (static member CanBind : ^Priority * ^TaskLike * (^TResult1 -> TaskStep<'TResult2, 'TOverall>) -> TaskStep<'TResult2, 'TOverall>) (x, y, __expand_continuation))
+
+    let inline RequireCanReturnFrom< ^Priority, ^TaskLike, 'T when (^Priority or ^TaskLike): (static member CanReturnFrom: ^Priority * ^TaskLike -> TaskStep<'T, 'T>)> (x: ^Priority) (y: ^TaskLike) = 
+        ((^Priority or ^TaskLike): (static member CanReturnFrom : ^Priority * ^TaskLike -> TaskStep<'T, 'T>) (x, y))
+
+    type TaskLikeBind<'TResult2> =
+        // We put the output generic parameter up here at the class level, so it doesn't get subject to
+        // inline rules. If we put it all in the inline function, then the compiler gets confused at the
+        // below and demands that the whole function either is limited to working with (x : obj), or must
+        // be inline itself.
+        //
+        // let yieldThenReturn (x : 'TResult2) =
+        //     task {
+        //         do! Task.Yield()
+        //         return x
+        //     }
+
+        static member inline GenericAwait< ^Awaitable, ^Awaiter, ^TResult1, 'TOverall
+                                            when ^Awaitable : (member GetAwaiter : unit -> ^Awaiter)
+                                            and ^Awaiter :> ICriticalNotifyCompletion 
+                                            and ^Awaiter : (member get_IsCompleted : unit -> bool)
+                                            and ^Awaiter : (member GetResult : unit -> ^TResult1) >
+            (awaitable : ^Awaitable, __expand_continuation : ^TResult1 -> TaskStep<'TResult2, 'TOverall>) : TaskStep<'TResult2, 'TOverall> =
+                let awaiter = (^Awaitable : (member GetAwaiter : unit -> ^Awaiter)(awaitable)) // get an awaiter from the awaitable
```

Comment:
I'm not sure if all awaiter implementations are immutable structs. Might want to mark it mutable.

C# stores local awaiter on statemachine after calling IsCompleted, allowing mutations during that call.
Note that I haven't actually seen a mutable awaiter ever, at least not in the BCL.

In any case it doesn't hurt to remove the copy for calling its methods.

Reply:
This awaiter goes on the state machine because it is used after `CONT`.  I've made it mutable, thanks!